src/read-file.c

```diff
@@ -18,56 +18,125 @@
 
 #include "read-file.h"
 
+#include <sys/stat.h>
+#include <stdio.h>
+#include <stdint.h>
 #include <stdlib.h>
 #include <string.h>
 #include <limits.h>
+#include <errno.h>
 
 char *
 fread_file (FILE *stream, size_t *length)
 {
-  char *buf = NULL, *tmpbuf = NULL;
-  size_t off = 0;
+  char *buf = NULL;
+  size_t alloc = BUFSIZ;
+  {
+    struct stat st;
+
+    if (fstat (fileno (stream), &st) >= 0 && S_ISREG (st.st_mode))
+      {
+        off_t pos = ftello (stream);
+
+        if (pos >= 0 && pos < st.st_size)
+          {
+            off_t alloc_off = st.st_size - pos;
+            if (SIZE_MAX - 1 < alloc_off)
+              {
+                errno = ENOMEM;
+                return NULL;
+              }
+
+            alloc = alloc_off + 1;
+          }
+      }
+  }
+
+  if (!(buf = malloc (alloc)))
+    return NULL;
+
+  {
+    size_t size = 0;
+    int save_errno;
+
+    for (;;)
+      {
+        size_t requested = alloc - size;
+        size_t count = fread (buf + size, 1, requested, stream);
+        size += count;
+
+        if (count != requested)
+          {
+            save_errno = errno;
+            if (ferror (stream))
+              break;
+
+            if (size < alloc - 1)
+              {
+                char *reduce_buf = realloc (buf, size + 1);
+                if (reduce_buf != NULL)
+                  buf = reduce_buf;
```

Comment:
what happens if realloc fails and `reduce_buf` is NULL?

Shouldn't we free (buf) and return NULL?

Reply:
Hey @giuseppe thanks for catching this. So I tried to treating `realloc` as an option. If it successfully allocates memory it'll reassign otherwise it will continue with the old assigned buffer. Idea was to `realloc` only if possible otherwise keep everything as it. But yeah nice catch we can also `free(buf)` and return `NULL` but that will fail the business logic wouldn't it. Let me know what do you think